shims/common/src/main/scala/com/intel/oap/sql/shims/SparkShims.scala

```diff
@@ -121,4 +121,20 @@ trait SparkShims {
   def getEndMapIndexOfCoalescedMapperPartitionSpec(spec: ShufflePartitionSpec): Int
 
   def getNumReducersOfCoalescedMapperPartitionSpec(spec: ShufflePartitionSpec): Int
+
+  def isLeafPlanExchange (plan: SparkPlan): Boolean = {
+    plan match {
+      case e: Exchange => true
+      case other => false
+    }
+  }
+
+  def sanityCheck(plan: SparkPlan): Boolean =
+    plan.logicalLink.isDefined
+
+  def supportAdaptive(plan: SparkPlan): Boolean
+
+  def supportAdaptiveWithExchangeConsidered(plan: SparkPlan): Boolean = {
+    isLeafPlanExchange(plan) || supportAdaptive(plan)
```

Comment:
It looks isLeafPlanExchange should not be checked recursively. So we moved it from #supportAdaptive.